### DISCUSSION TITLE
Refactor the AaC language API

### DISCUSSION THREAD

**Discussion by jondavid-black** [DevOps-MBSE/AaC#691] [Ideas]
I've been working through what turns out to be some pretty significant changes lately and it has been pretty frustrating.  The root of that frustration is that I find it difficult to work with the core API of the language itself.  Having a definition is useful, but it remains difficult to navigate within the definition.  I'm contemplating a little experiment...with no real understanding of if/how it may work out.  What if the core API was generated from the core language specification in "plain old python"?  Could we create a minimal set of utility implementation that would provide a CLI and gen-aac type behavior that we build everything from?

So when I think of the language API, I think of 2 key things...the type (aka Schema) and the instance (aka Definition).  Types tend to be things with a root key of schema, other than enums.  Instances tend to be things with a root key other than schema.

Another challenge I'm struggling with is probably just a style thing.  I'm and old school OOA/OOD bubba, and get lost with all the functional programming style stuff we've got in place.  It's not wrong, just difficult for someone without extensive background working in this style.  As we consider implementation of inheritance, I fear managing a fundamentally OO model using a functional style will be even more challenging to navigate.

So to clarify, the hypothesis is that working with AaC could be easier if we generate the AaC API from the AaC core model and provide a more "natural" API that aligns directly with the language definition in the core spec.

To test the hypothesis I propose a fork of the repo to keep everything clean.  We would modify the CLI to understand a single "hand jammed" command for ```gen-aac``` that would generate python code based on the AaC core spec.  Another "hand jammed" utility would be a revised ```LanguageContext``` that would keep track of everything.  Anything that has a ```schema``` root key would have a python class generated for it.  Inheritance would be implemented using python's class inheritance.  Instances would be instances of the python class. The ```LanguageContext``` would keep a list of instances (aka Definition today) and a list of types (aka Schema instances).  As a plugin-developer, you would get the instance of the type you want with a given attribute like name.

```python
my_plugins = language_context.get_instances_of_type("Plugin")
```

To work with an instance, you just use standard python syntax.

```python
some_plugin = language_context.get_instance_of_type_by_name("Plugin", "CoolStuff")
if some_plugin:
   for command in some_plugin.commands:
      print(f"Plugin {some_plugin.name} has command {command.name}")
```

If you want to understand if an instance is of a certain type you just use python's built-in ```isinstance``` function.

I don't have my head around enums and primitive types.  There's some thought to work through on that front.  Enums would hopefully just use the python Enum approach.  Perhaps we just require that any AaC primitive definition be assigned a python primitive (integers, floats, booleans and strings) to enable code generation.

As we look at some of the upcoming work, we're going to risk fundamentally writing a whole new language infrastructure.  I think we can avoid that and simplify the extensibility of AaC all at once.  Of course, all the plugins we have today would have to be rewritten...which is unfortunate...but I think would be less work than building out a bespoke language infrastructure and then rewriting the plugins for that.

There's a lot to think about here.  That said, perhaps it wouldn't be too difficult to build out an initial prototype with only the core language content.  Any additional thoughts or concerns would be welcome!

**Comment by jondavid-black** [comment]
Update on the experiment mentioned above:

I've begun work on the pure python API.  Let's call it the Plain Old Python (POP) API.  

Everything is sell self-defining.  The aac language definition itself is a collection of schemas, enums, and custom root items that will produce python classes.  The LanguageContext has been updated to parse an AaC file into a definition, preserving access to the dict, and then instantiates python objects that can be access and used "natively" by plugin developers.

This has pretty much turned into a rewrite.  The core functionality seems to be working.  Up to this point, the code has been hand crafted with the intention of taking the result and folding it into the gen-plugin templates.  At this point, gen-plugin can generate a working plugin.  Next steps are to autogenerate projects, schemas, enums, etc.

Since this is a rewrite, a handful of things are being changes.  First off, the generation plugins have been tweaked.  There is now a generate plugin that uses a generator schema to define its behavior.  Rather than custom building dicts to pass into jinja, it just passes in an AaC model dict (aka Definition.structure).  This means that if you understand the AaC model, your jinja template is pretty understandable.  I've also created a concept of ```run_before``` and ```run_after``` for plugins.  The idea here is to declaratively establish cross-plugin relationships.  This is currently used to show that gen-plugin uses generate.  In fact, gen-plugin has very little implementation beyond defining the generators and templates to be used by the generate plugin.  Once the ```check``` plugin is in place, I anticipate it being used as a ```run_before``` reference rather frequently.  This is currently working, but there may be a better API that could be created to help manage plugin cross-reference execution.  We'll leave that open for future discussion.

Other changes:
  - Primitives are no longer a "magic enum".  Define a primitive using the ```primitive``` root.
  - Inheritance is in place in the model.  You can now extend schema, enum, and primitive.  There is also an abstract type called ```AacType``` that each of these inherit from.  This will allow a user to more generalized in referencing types within their model for situations where they may not know if the reference is for a primitive or schema.  Extending ```AacType``` should be extremely limited, but the capability is there if needed.
  - Modifiers are in place.  For now, we're just using ```abstract``` and ```final``` to characterize inheritance.
  - References are in place.  A reference definition takes a "fully-qualified model path" and expects to find a matching item.  For example if you're referencing a schema by name the reference would be ```reference(schema.name``` and the entry in the AaC model for that reference field would be the name of the schema.
  - Packages are in place.  Not having packages was creating obstacles for definition and generation.
  - Tweaks have been made to plugin execution and the CLI.  This was just a refactor to simplify plugin registration and execution.
  - The AaC package structure has changed a bit.  All things "contextual" are not in the ```aac.context``` package, including ```LanguageContext``` which is now implemented as a self-contained singleton.  Other contextual items like ```Definition```, ```Lexeme```, and ```SourceLocation``` are here as well.
  - The ```aac.cli``` package has transitioned to the ```aac.execute```.  This new package manages all execution details, including CLI and plugin execution.
  - The ```aac.io``` package remains in place and is largely unchanged.
  - The ```aac.lang``` package now contains the AaC self-defining model and the generated core language python classes.
  - The ```aac.plugins``` package is now just a holder for bult-in plugins.  All plugin execution and registration has been transitioned out.  There's also no concept of first-vs-third party plugins anymore.  The generate function now just asks where you want to output code, tests, and docs and behaves accordingly.  There is also now a way to model an AaC project similar to how we would model an AaC plugin, so project generation is now explicit rather than inferred.

While there's still a lot to learn as we progress on this experiment, I'm very excited with the intermediate results.  Working with simple python constructs to navigate and utilize model content has been very nice.  The various tweaks that have been put in should also accelerate plugin development and language extension (i.e. SysML, UAF implementation).

With the core language now implemented by hand and basic plugin generation working, I'm now diving into schema and enum generation.  My intention is to completely replace all hand-written core model code with generated code.  More to follow.

All the updated code is in the pure-py-api branch.  There's still quite a bit of work do to, particularly with unit testing (I have not been good about that one).

**Comment by jondavid-black** [comment]
Just a quick update...

The chicken-and-egg problem is now behind me.  All the hand-generated code put in place to make everything work out of the gates has now been completely blown away and replaced by output of gen-plugin using the aac.aac language definition.  We are officially self-defining under the new POP API.  There's still a lot of missing definitions, but now they can all be added by simply updating the language definition and re-running gen-plugin!

**Reply by jondavid-black** [reply]
Well, I may have gotten a little too excited.  I now realize that generating new components within the aac language definition isn't quite as trivial as I had hoped.  This is caused by the LanguageContext loading the aac.aac file when it initializes the singlton.  Since it sees new things that don't have generated code, it blows up when trying to load everything.  But there is a workflow that lets you work around this limitation...and I think this should only ever apply when you're working in the aac.aac file directly.

- Write up the new language definitions in a separate aac file.
- Run gen-plugin on that new file.
- Copy the language definitions to the aac.aac file once there are python objects for them.
- Rerun the gen-plugin to make sure everything is happy.
- Run clean
- Delete the separate aac file.
- Done!

**Reply by lizzcondrey** [reply]
I am hoping to get some time later this week to try and get familiar with the branch and the changes outlined in above comments. Definitely sounds like this has been an enlightening and fruitful experiment. I am intrigued to dig in and play around with it and see what snags I hit. Thanks so much for the updates and captures of the snags and resolutions!

**Comment by jondavid-black** [comment]
All previous AaC core schemas, enums, and primitives are now in place, but constraints are not yet defined or applied.

Also defined and implemented an updated version of the ```print-spec``` plugin which is now called ```print-defs``` and gives the user the option to print all definitions (default) or only the "core" definitions by providing the ```core-only``` parameter.

Next steps:
1. Implement new jinja templates for constraint generation through gen-plugin
1. Implement at least 1 ```context_constraint```
1. Implement at least 1 ```schema_constraint```
1. Implement at least 1 ```primitive_constraint```
1. Build the ```check``` plugin to run the constraints
1. Setup any appropriate ```run_before``` plugin updates to include ```check```
1. Consider adding a check cache to prevent redundant execution for same outputs that could occur with ```check``` declarations across plugins.
1. Implement remaining ```context_constraint```
1. Implement remaining ```schema_constraint```
1. Implement remaining ```primitive_constraint```
1. Ensure all constraints are properly applied

Once this is all complete, it may be time to put up a PR.

**Reply by jondavid-black** [reply]
FYI...I do realize there is a documentation update to be done.  Personally, I'd like to get the PR started to initiate peer-review and work documentation in parallel.

**Comment by jondavid-black** [comment]
Another quick progress update...

I completed: 
- The jinja templates and code generation for ```context_constraint```, ```schema_constraint```, and ```primitive_constraint```.  I've actually been very happy with the updated code generation.  It now just uses the AaC definition without a bunch of "extra magic" in the dict that feeds jinja.  At least for me, this makes it all much easier to comprehend.  That said, where there are little nuggets of "magic" needed, I've created the ability to define python helper functions and feed them into the jinja template.  I've purposefully kept these trivial to avoid adding unwanted complexity...but there's really no limit to what could be done behind the covers using these helper functions.
- The ```check``` command is now built and will execute all constraints correctly against the AaC language definition
- All the primitive constraints are in place
- Changed the reference primitive into separate ```dataref``` and ```typeref``` because the underlying meaning and constraint logic for those is different.  I think the new approach is much clearer...although references in general are still a bit complex and will need good user documentation.
- Along with the reference changes for ```dataref```, it is now possible to find fields using a "field chain" (i.e. ```plugin.commands.acceptance```) and the constraint will "walk the tree" to verify everything is in place.  This may be useful to expand into a reporting function in the future.  I wish I could come up with a better name for this.
- ```reuired_fields``` schema constraint is in place
- ```no_undefined_fields``` schema constraint is in place
- ```unique_root_keys``` context schema constraint is in place
- Tweaked he ```print-spec``` command to be ```print-defs``` that gives the user an option to only do he core AaC lang or do all known definitions
- Added code generation for gherkin feature files (this was pretty cool, not generate can use a sub-set of a definition based on the template definition)
- I went ahead and created a new project for the AaC VSCode plugin.  The intent is to get the "core" AaC project as clean as possible in the near future.  I'm not sure if we want to go ahead and make this update now or wait...but if we don't have active VSCode users right now I'd say we should jump on the change.  I'd also move all the LSP stuff over there and keep it located together.

I haven't:
 - Setup ```run_before``` to do ```check``` prior to ```generate``` or ```gen-plugin```.
 - I really want to consider some sort of "run cache" to prevent the same command being executed multiple times for the same inputs due to the ```run_before``` and ```run_after``` chaining
 - Reviewed other "validations" from the current baseline and see what remains to be transitioned
 - Started reworking documentation...but I keep thinking about it.  

Some bugs and tasks I've noted:
- Bug:  Currently the plugin generation is hard coded to expect your ```.aac``` file to have the same name as the plugin.  It isn't a big deal (easy fix) but may be a frustration for users trying to extend AaC in the future.
- Bug:  If previous gen-plugin resulted in a python error, the next run fails when trying to load the command on launch.  This can be easily corrected by deleting the bad files from the previous gen-plugin run.  This may have just been an artifact of working through the code generation templates and logic and may not persist once we're satisfied things work properly.
- Bug:  The ```from_dict``` approach to generating immutable objects works for the local fields, but "hopes for the best" on inherited fields.  Need to figure out some way to properly handle ```super``` construction for sub-types.  May have to move away from immutable types to make this feasible.  One option I'm considering is to segregate the population of ```args``` and ```kwargs``` into a separate function that could be used by child classes as well as the ```from_dict``` call.  In theory this should be a pretty simple change...but will have to burn a few more grey cells to know for sure.
- Task: Figure out how to add file location data to constraint results.  I've dropped in the definition source and lexemes, but lexemes are still a mystery to me and I don't understand how to find specific information from them yet.  Not sure if this is easy or hard, just need to research the lexemes and use them correctly. Maybe we just need some helper functions to simplify this?
- Task:  When the generator asks you if you agree with the output paths, it only includes the top level info.  It would be better to include the full path if possible.  The rub is the package name may not be universal so this could result in quite a few entries for the user to review and approve.
- Task:  I'd really like to add a ```universal``` boolean to the schema_constraint definition so that the user doesn't have to include the constraint explicitly on all schemas.  This is just a usability feature, it works now but requires explicit constraint allocation for all schema_constraints.  It would have to be used judiciously, but things like ```required_fields``` are pretty universal in nature now that the ```is_required``` attribute is included in the field definition itself rather than a separate constraint argument.
- Task:  Make sure polymorphism is in place and working.  The updates that are now in should make this easier, but it will probably still be a brain twister and need some solid testing.
- Task:  I'd like to tweak the use case to better align with SysML out of the gate.  I've been reviewing the SysML spec and usage in other tools.  It actually looks like a simple tweak of the schema definitions and make make use cases more straight forward for users.

I'm intentionally putting off:
- Anything related to puml...want to get this into a 3rd party plugin once we publish the new release.
- Anything related to protobuf...want to get this into a 3rd party plugin once we publish the new release.
- Needed updates to the 3rd party AaC-BOM plugin...waiting for the new API.
- Anything releated to the REST API...want to get this into a 3rd party plugin once we publish the new release.
- Updates for the ```gen_design_doc``` plugin...just because I want to contemplate this one a bit more.  We may be better off providing users with a more flexible approach that leverages the new ```generate``` command along with their own custom templates.  More to think / talk through on this one.  But as I write this I'm realizing it could be cool to create a custom document generator to use as part of our own document generation....maybe prove it's is reasonable and valuable to use before expecting others to try it.
- The ```help-dump``` command...mostly because I don't understand how a JSON dump of the CLI help is useful to end users at this point.
- The ```gen-json``` command...mostly because I'm not sure it provides user value.  After all, it was originally just a way to test commands with the simplest possible logic I could think of.
- Anything related to plugin management...because I'm not sure it's needed if we move stuff into 3rd party plugins.  Let ```pip``` be your plugin manager!

So what is really needed to get to a PR?  I think we need a good run at updated documentation.  We don't want to risk overwhelming user confusion (especially now that things aren't just public but "known").  I also want to beef up testing quite a bit.  Most of the ```io``` and ```context``` tests just need to be transitioned over from the prior code base...with little change.  The big one I feel is needed is to add unit tests for plugins and constraints.  And of course, the old acceptance tests still need to be transitioned into the baseline and re-ran through gen-plugin to produce the feature files.  I'd love it if we could get a set of working acceptance tests baked into the pipeline!  All in all, I don't think there's a huge amount left to complete before we're "PR Ready", but I don't want to get overaggressive either.

More to follow!

**Comment by jondavid-black** [comment]
Time for another update!  It's hard to believe this little adventure started almost a month ago.  The whole experience has been really enjoyable...even with some real head scratchers popping up along the way.

I completed:
- Added the ```run-before``` to ```gen-plugin``` so invalid input will now prevent attempts at generation.
- Added unit test generation for plugins to include commands and constraints.
- For commands, unit tests can either ben "internal" or "CLI execution" based.
- Regenerated and wrote unit tests for ```version```
- Regenerated and wrote unit tests for ```aac_primitives```
- Reviewed prior plugins / validators and down-selected items for this version.  I'll go into more on this below.

I haven't:
- Finished regenerating and writing unit tests for all plugins...but it turns out ```gen-plugin``` was somewhat "all encompassing" due to the ```run_before``` invocation of ```check``` and the ```run_after``` invocation of ```generate```, so I'm hopeful the rest will be a relatively light lift.
- Built unit tests for ```aac.context```, ```aac.execute```, and ```aac.io```, which are the only non-generated portions of the aac code base.  I'm expecting to just copy many of these from the previous baseline.
- Refactored the VS Code extension and related python plugins to a separate repo.  I really want to get this code out of the core repo so that we can get back to a "pristine" depend-a-bot state by removing the javascript stuff.
- Clean up the original code directories that I've kept around for quick reference...although I've mostly used GitHub's excellent code navigation functionality when needed.
- Reviewed and updated the documentation.  This new internal API does have some real-world impacts on the workflow to be used by developers who would want to extend aac.  There are also some insights on how things like the ```check``` command invokes ```constraint``` items defined in plugins that needs to be clearly described for aac extension developers.

Plugin Overview:
- ```version```: in place and working
- ```validate```: replaced by ```check``` which is in place and working but still needs unit tests 
- ```specifications```: omitted for now, but will plan to implement in a future release
- ```active_context```:  eliminated the whole concept of plugin management, and therefore this plugin
- ```gen_design_doc```:  omitted for now, but may revisit later with an understanding that this should be AaC specific and other users would almost certainly choose to implement their own document generation for their unique context
- ```gen_gherkin_behaviors```:  omitted as a stand-alone command and integrated into gen-plugin, but may revisit for a future release
- ```gen_json```:  eliminated because I don't think it had any value to anyone...this was just a test case for when we initially tried to create commands to prove we could do something relatively simple
- ```gen_plant_uml```:  eliminated with the intention of refactoring out into a [3rd party plugin](https://github.com/DevOps-MBSE/aac-puml)
- ```gen_plugin```:  in place and working
- ```gen_protobuf```:  eliminated with the intention of refactoring out into a [3rd party plugin](https://github.com/DevOps-MBSE/aac-protobuf)
- ```help_dump```:  eliminated because it isn't clear how a json dump of the CLI help is useful to users, but may revisit if we discover a use case in the future
- ```lsp_server```:  eliminated with the intention of refactoring out into a [3rd party plugin](https://github.com/DevOps-MBSE/AaC-VSCode) along with the VS Code extension
- ```plugin_management```:  eliminated the whole concept of plugin management, and therefore this plugin
- ```primitive_type_check```:  in place and working as ```aac_primitives```
- ```print_spec```:  replaced with ```print_defs```
- ```rest_api```:  eliminated for now but may revive later as a 3rd party plugin
- ```defined_references```:  replaced with primitive constraints
- ```exclusive_fields```:  in place and working, but needs unit tests
- ```reference_fields```:  replaced with primitive constraints
- ```reference_targets```: replaced with primitive constraints
- ```required_fields```:  in place and working, but needs unit tests
- ```requirement_references```:  no longer needed due to how reference primitives work now
- ```root_keys```:  replaced by ```unique_root_keys```
- ```subcomponent_type```:  no longer needed due to how reference primitives work now
- ```unique_names```:  omitted for now, but may need to revisit if we encounter a case where the new package and name construct doesn't serve the purpose
- ```unique_requirement_ids```:  omitted for now, but may add in a future release...or even this one depending on urgency
- ```unused_definitions```:  omitted for now, but may add in a future release if it doesn't interfere with the AaC developer workflow using the new API
- ```usecase_participants```:  omitted for now, but top of the list to roll back in due to recent user feedback on use cases...and an accompanying update to better align use cases with SysML utilization
- ```validator_implementation```:  this is all rolled up in the "normal" constraints and language context now...nothing special

There are a few new plugins to mention:
- ```constraint_assignment_arguments```:  This was the result of one of those "head scratcher" moments mentioned above.  With the new API I wanted a simple and clean way to define constraint arguments in the constraint definition as fields, but when assigning the constraint there was no good way to have the values align with the field definition.  Hence the birth of this constraint and the ```any``` primitive type.  This allows named parameters to be declared in a constraint assignment and ensures they are consistent with the argument fields in the constraint definition.
- ```generate```:  This is a generalization of the previous ```gen_plugin```.  The AaC language now provides a highly flexible means of defining generation of pretty much anything using this command.
- ```if_true_then_empty```:  This constraint arose from the creation of the ```universal``` flag on ```schema_constraint```.  Since the ```universal``` flag provides a shortcut to run the schema constraint on everything without having to do an explicit assignment, any flagged schema constraint would have to be defined with no arguments.  This just provides a relatively generic way to enforce that logic.

So that's a summary of where we are with the Plain-Ole-Python (PoP) API work.  It's been a pretty heavy lift so far, but when I actually use the new API to implement functionality it has been a great improvement over the old dict-based strategy.  The AaC developer workflow does have to change a bit, but I believe we've managed to keep the flexibility while improving approachability, usability, and extensibility.  I'm hopeful to get through these last bits over the next few days to the point it would make sense to publish a PR for everyone's feedback.

**Reply by lizzcondrey** [reply]
Wow, you've really covered some ground again in this most recent stint. I think I've followed everything, but I will give it a read again hopefully tomorrow to post back some questions to make sure I'm tracking everything correctly and that we're prepared for the PR.

Glad to hear it's been enjoyable and have had a decent amount of "aha" moments it sounds like with developing the new API and infrastructure. I'm definitely looking forward to evaluating and seeing if this resolves the handful of pain points experienced in my short time on this project.

Thanks so much for keeping a solid record and update thread of your progress!

**Comment by jondavid-black** [comment]
One more quick update...

I've been in test-mode for a little while and now have unit tests over the 80% mark when run through tox!  Definitely hit some snags along the way...and my old java habits had to change to get over them.   But all good...things are on track for a PR in the relatively near future.

I have taken a few precautions, just to give us a "safety net" with such a large change.
- I forked the current AaC project.  There are a handful of items mentioned above that I'm intentionally omitting so that we can transition them to plugins in separate repos.  This will give us an "easy access" view of all the pre-new-API stuff.
- I went ahead and transferred the VS_Code extension and LSP plugin over to the new AaC-VSCode repo.  I fixed the various imports, but there's still a lot of "old API" stuff that needs to be updated.  I think we can just tackle this incrementally after we get the new API in place.  It's just a bit too hard to work the API changes across 2 repos simultaneously at the moment.  Besides, it's a lot of work too and I don't want to allow perfection to become our enemy.  That said, I did use the gen-project and gen-plugin commands from 0.4.0 to recreate things in the new repo and it worked fine!  The code is copied over and just need to finish the refactor.
- I have removed the workflows for the VS Code build in the new branch...but they're still visible in the fork if we need them again.

At this point I'm down to a few final plugin test implementations and documentation updates.  I've also got a note to double-check cross-platform stuff to make sure I've not gone too "linux first" with things.

Will keep plugging...look for a PR in the not-too-distant future!

**Comment by jondavid-black** [comment]
Another update...and a request for opinion.

I've continued to focus on testing, and that has been productive.  I've discovered and corrected many defects along the way, particularly for the check plugin and various constraints.  The details aren't too important for discussion, other than to say I'm just about done with testing but have a few items I'm considering before claiming success.

1.  Automate Acceptance Testing:  The ```python/model``` folder has been a little playground that I've used to help me think through model definition changes and support testing.  That said, the testing has always been manual...and therefore not particularly well maintained.  As I've worked through updates to ```python/model``` for the 0.4.0 baseline I've actually had some good discoveries and bug fixes that "normal unit testing" just didn't uncover.  So I think it would be valuable to transition the ```python/model``` folder to an automated acceptance test using [Behave](https://github.com/behave/behave).  The thought here is to keep things as simple as possible.  In pretty much every case, the manual testing is little more than running ```check``` against a defined AaC model.  I think this can be setup so that a consistent set of parameterized steps can be used for various feature files that each point to different model definitions.  I'd like to get that automation in place and then swizzle it into our tox execution.
2. As I've been testing I've encountered a point of frustration in the AaC workflow that I feel needs to be resolved.  The current "Pure Old Python" API is built around code generation.  If you need some new type, you define the Schema for it, run ```gen-plugin```, and then you can create instances of that Schema in your model.  Seems simple enough, but there are cases when we will want to define the Schema and an instance together in the same model.  The current workaround it to delay defining instances until after you've run ```gen-plugin``` but I don't think that's sufficiently "usable" for our target audience.  I'd like to explore eliminating the code-generation step for Schemas used within AaC and replace it with dynamic python class generation.  After a little examination, I believe this change would be a relatively contained update to ```LanguageContext``` and deletion of some stuff from ```aac/plugins/generate```.  Here's where some feedback would be valuable.  Should we just stick with the current solution and workflow for ```0.4.0``` or go ahead and make the change?  I'm leaning toward making the change, but I'll wait a day before jumping in on that in case anyone has other thoughts.
3. Generating Acceptance Tests:  We currently generate feature files for our plugins, but we don't really do anything with them.  I'm considering adding code generation to produce stubs of the steps for automated acceptance execution using Behave.  This isn't something that I'd do until finishing up 1 above, and it could be deferred for now.  Should this be included in the ```0.4.0``` release or should we wait?
4. And now for the "really big one".  I've gone in and updated the ```usecase``` to align with SysML.  As I've been looking deeper at SysML, I'm starting to think maybe we should just go "all in" on SysML as our "first class" modeling language rather than as a 3rd party plugin.  Now to be clear, I'm not suggesting we put in every little thing up front, but I think there are a few simple changes that would set us down this path.  The most obvious change would be to rename ```Model``` to ```Block``` and tweak the ```state``` and ```behavior``` a bit to align better with the SysML spec.  I think we'd want to update to use SysML interface definitions as well, which should accommodate the functionality we previously discussed for ```Channel```.   I'd recommend we keep ```Schema``` for now.  I know I've been negative on this in the past, but the more I think about it the more I think using "SysML vocabulary" would actually improve approachability for our users.  I'm really thinking this would be more of a ```0.5.0``` release, but something worth discussion in the near term.

Those 3 (+1) items are the only real "big changes" remaining, but there's still some work to finish up.
1. Finish writing unit tests for plugins and constraints.
2. Make sure there aren't any "os-specific" things that need to be corrected.
3. Update docs
4. Expand acceptance tests a bit (if there's something we don't feel is sufficiently covered by existing items in ```python/models```.

**Reply by jondavid-black** [reply]
A little more thought on SysML as the "first class" modeling language.  I'm not sure there's a lot of practical advantage to that change, but I think the approachability benefits may be pretty substantial.  There are people who have spent a considerable amount of time, energy, and money becoming proficient in SysML as part of their MBSE journey.  I think aligning to that would provide them value.  Perhaps simplified model language constructs could be applied on top of that for other use cases.  In the end, I'm still bouncing back and forth on the idea and probably need some encouragement one way or the other.  That said...a "SysML compliant" badge on the home page would be a good thing.

**Comment by jondavid-black** [comment]
It's been a couple days, so here's another update.

I just finished (well, made work anyway) the updates associated with "big change 2" from the last post.  I've removed code generation for AaC definitions and now generate everything on-the-fly.  This eliminates the whole "chicken-and-egg" issue with code generation.  Now the LanguageContext just dynamically walks through the definition heirarchy, inheritance, and all that junk to just create what's needed when it is loaded.  It also ensures everything that's loaded gets a python object populated in the instance field of the definition.  There's also now the ability to define types as just names (as has been done historically) or as fully qualified names using the package and name.  If there happens to be a conflict when just using the name, it will throw an exception and tell you what the conflicting items are.

Anyway...that all sounds great, so what's the catch?  Plugin generation has to change a bit.  Since there are no longer python files that actually define the AaC types, you can't import them in the plugin file or use typing to convey function parameters for content dynamically created from AaC definitions.  My current thought is that this is an acceptable trade off.  My rationale is simply the fact that python is a dynamically typed language, so this isn't completely foreign within this environment...although as an old java/c++ developer with a preference for static typing it isn't my favorite outcome.

So to complete the updates for "big change 2" I really just need to update the generator to not insert imports or parameter types for AaC definition types.  Should be pretty quick update.  This did expose my lack of real testing though.  I was passing the tox test coverage limit because the generated code had generated tests that was artificially driving up the coverage averages.  Now I really have to buckle down and implement plugin unit tests...as mentioned before.

Once I get this portion done I'm going to jump onto "big change 1 and 3" and fully embrace automated acceptance testing.

FYI...I haven't pushed this change into the branch yet.  I really want to clean up gen-plugin first because as it stands it will produce unusable code.  Hope to get that wrapped up today now that I'm not tied up in all-day meetings.

**Reply by jondavid-black** [reply]
OK, commit and push has been made.  Tests are passing but I still need to finish writing tests for plugins.

**Reply by jondavid-black** [reply]
And now the commit for "Big change 1" is in place.  It was actually pretty easy to define the feature files and step impls...very little code.  Having done this, I'm not completely sure if we truly need to do "Big change 3" other than to recognize there is a learning curve to consider for someone who's never defined automated acceptance tests before.  The real realization, and the thing that kept it all very simple for me, was to provide step consistency by using step parameters.  I'm just not sure we really need to drive that sort of detail into the baseline right now.  Feels like an improvement for later.  Of course, some of that feeling is the realization that this is rapidly becoming a long lived branch and I'd really like to wrap it up ASAP.

**Reply by jondavid-black** [reply]
And now behave will be executed when we test with tox.

**Reply by jondavid-black** [reply]
OS-specific things (i.e. "\n" in strings) has now been replaced by the os.linesep OS-independent solution.  Only a few more steps to go to achieve "PR ready".

**Comment by jondavid-black** [comment]
It's been a "docs-or-bust" sort of day.  I've updated the project documentation to provide information on the new language API, plugin generation, and relevant home page updates.  I've removed a handful of docs that are no longer relevant with the new core API and associated language model changes.  Of course, I had to provide examples in the documentation which made me realize that now I have to actually code up those examples and create tests for them.  It would be too embarrassing if an example didn't actually work.

I still need to work through the user guide and then actually test the document generation, but it feels like there was some good progress on some work I really wasn't looking forward to (although, this was so much better than building Word or PowerPoint documentation).

I believe I'm down to finishing up an handful of automated tests (including creating tests for examples used in the docs) and then it will be time for a PR...finally!!!

**Comment by jondavid-black** [comment]
It's finally done..and PR #714 is up for review.

All in all I'm pretty pleased with how it turned out.  The handful of plugins I made as part of the new API were actually rather simple and quick, so mission accomplished in my book.  That said, there are a few things I'd still like to tweak once we're past this PR.

## Packages

The more I think about it, the more I think there should be a separate package construct.  This is nothing new...UML and SysML do this as well.  When we originally put AaC together we were thinking purely in terms of modeling the decomposition of a system aligned with traditional system engineering and spec tree concepts.  So in that mindset we abstracted models and packages into a single thing.  This is why `Model` has the `components` field...so that it can be both a "thing" and a "container".  At this point, I think there's value in separating "things" and "containers".    It would probably look something like this:
```yaml
schema:
  name: Package
  description:  A package serves as a container that enables logical grouping of model components.
  root: package
  fields:
    - name: name
      type:  string
      description:  The name of the package
    - name: description
      type: string
      description: The description of the package
```
And then other items would be able to reference the package in their declarations using a dataref(package.name) field type.

## Context Generate

The `generate` plugin is now setup to assume there's a single AaC model definition that drives content generation.  For the cases we've needed this has been sufficient, but I can imagine cases where we need something more.  For example, let's assume we want to create a package diagram with classes...using AaC `package`, `model`, and `schema` types.  Assuming the package diagram is a single content item, we don't have a good way to handle this. 

So let me just emphasize, I've very happy with the simplification we put into `generate` and am not recommending a change.  I'm just thinking of the next step where some "meta-model" is created and passed into the generator.  I'm not 100% certain I could articulate how to set that up in a reasonable manner, but I think it would be a valuable capability.

## Plugin Structure

While I'm not upset with the current structure, if we end up adding more core plugins, things could be a little messy.  If it gets to that, I'd suggest we add `commands` and `constraints` as sub-modules under `plugins` to keep things a bit more tidy.

## Reporting

Although we've got some of the core pieces in place with the `dataref` type, it would be great to think through how we could create a general way to aggregate and filter data for reporting.  I can see this being a key enabler to the "meta-model" concept described above.  I think one approach to this could be something similar to how `dataref` works using the field-chain concept but would need to be augmented with a filter.  We may also need to flatten the data, but that's worth some deeper thought as a structure could be valuable.  Anyway, this is just a future thought but one that could be valuable in a number of use cases.

## Conclusion

This whole experience has been a lot of fun.  There was a ton of back-and-forth with certain design concepts, but I'm pleased with how it ended up.  I think the new API really simplifies extensibility...which was the primary goal.  I think the updates from the old "validator" scheme to the new "constraint" scheme is a substantial improvement as well.  I'm very happy that we were able to remove so much code from the base repo, but I also recognize there's a lot of work to be done to update the refactored stuff to align with the new API.  I'm very happy with the simplified execution structure.  I'm also very happy with the updated plugin definition and `gen-plugin` that (at least for me so far) really simplifies adding new behavior.  As I mentioned above, I'd still like us to consider adoption of a "standard" modeling language as our "first class" model solution because I think this would help encourage user adoption and usability.  While AaC is purposefully simple, I fear that the simple fact it is different may drive a perception of complexity.  Regardless, this little adventure has been a truly enjoyable experience.  This was the first time I've ever fully embraced a generative AI assistant while working on anything of any scale and complexity (granted, AaC isn't that large or complex) and to say it was eye-opening is an understatement.  I'm very much looking forward to merging and closing this PR and getting started on the next round of enhancements.
